### PR TITLE
Feature/add insert select statement in querybuilder

### PR DESCRIPTION
**Pull request by ghost**
related to issue #1811

This PR qill enable the posibility to create INSERT INTO SELECT Statements
examples:
1.
```php
$qb = $conn->createQueryBuilder()`
`->insert('users')`
`->select('u2.name', 'u2.password')`
`->from('users', 'u2')`
`->where('u2.id = ?')
```

2.
```php
$qb = $conn->createQueryBuilder()`
`->insert('users')`
`->values( array( 'name' => null, 'password' => null ) )`
`->select('u2.name', 'u2.password')`
`->from('users', 'u2')`
`->where('u2.id = ?')
```

### CODE REVIEW

File: tests/Doctrine/Tests/DBAL/Query/QueryBuilderTest.php [hunk 1]
```diff
@@ -476,6 +476,38 @@ public function testInsertValues()
...
+
+        self::assertEquals(QueryBuilder::INSERT, $qb->getType());
+        self::assertEquals('INSERT INTO users (SELECT u2.name, u2.password FROM users u2 WHERE u2.id = ?)', (string) $qb);
+    }
+
+    public function testInsertValuesSelect()
+    {
+        $qb = new QueryBuilder($this->conn);
+        $qb->insert('users')
+            ->values(
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Query/QueryBuilderTest.php]: This syntax is a bit weird. As far as I understand the resulting SQL, this method defines the columns which the selected columns will be mapped to. Not values.

File: tests/Doctrine/Tests/DBAL/Query/QueryBuilderTest.php [hunk 2]
```diff
@@ -476,6 +476,38 @@ public function testInsertValues()
...
+                array(
+                    'name' => null,
+                    'password' => null
```
Comment (morozov) [inline tests/Doctrine/Tests/DBAL/Query/QueryBuilderTest.php]: This is also weird. Why should there be nulls on the right-hand side?

File: lib/Doctrine/DBAL/Query/QueryBuilder.php [hunk 3]
```diff
@@ -1180,12 +1210,22 @@ private function isLimitQuery()
...
-        ' VALUES(' . implode(', ', $this->sqlParts['values']) . ')';
+        $statement = 'INSERT INTO ' . $this->sqlParts['into']['table'];
+        if( !empty($this->sqlParts['values']) && count( array_keys($this->sqlParts['values']) ) > 0 ) {
+            $statement .= ' (' . implode(', ', array_keys($this->sqlParts['values'])) . ')';
+        }
+
+        if( !empty($this->sqlParts['select'] ) ) {
+            $statement .= ' (' . $this->getSQLForSelect() . ')';
+        }
+        else {
```
Comment (morozov) [inline lib/Doctrine/DBAL/Query/QueryBuilder.php]: What if both `'select'` and `'values'` parts are defined?

File: lib/Doctrine/DBAL/Query/QueryBuilder.php [hunk 4]
```diff
@@ -1180,12 +1210,22 @@ private function isLimitQuery()
...
      * @return string
+     * @throws QueryException
      */
     private function getSQLForInsert()
     {
-        return 'INSERT INTO ' . $this->sqlParts['from']['table'] .
-        ' (' . implode(', ', array_keys($this->sqlParts['values'])) . ')' .
-        ' VALUES(' . implode(', ', $this->sqlParts['values']) . ')';
+        $statement = 'INSERT INTO ' . $this->sqlParts['into']['table'];
+        if( !empty($this->sqlParts['values']) && count( array_keys($this->sqlParts['values']) ) > 0 ) {
```
Comment (morozov) [inline lib/Doctrine/DBAL/Query/QueryBuilder.php]: Isn't it the same as `if (count($this->sqlParts['values']) > 0)`?

File: lib/Doctrine/DBAL/Query/QueryBuilder.php [hunk 5]
```diff
@@ -1180,12 +1210,22 @@ private function isLimitQuery()
...
+            $statement .= ' (' . implode(', ', array_keys($this->sqlParts['values'])) . ')';
+        }
+
+        if( !empty($this->sqlParts['select'] ) ) {
+            $statement .= ' (' . $this->getSQLForSelect() . ')';
```
Comment (morozov) [inline lib/Doctrine/DBAL/Query/QueryBuilder.php]: The parentheses around `SELECT` are not needed. At least in most cases.

Comment (viosys) [reply lib/Doctrine/DBAL/Query/QueryBuilder.php]: I know, but if it's a slightly more complex statement then I'm on the safe side

File: lib/Doctrine/DBAL/Query/QueryBuilder.php (hunk 3 above)
Comment (viosys) [reply lib/Doctrine/DBAL/Query/QueryBuilder.php]: then a 'normal' inster statement is generated. You might even throw an exception if both are specified

File: lib/Doctrine/DBAL/Query/QueryBuilder.php (hunk 4 above)
Comment (viosys) [reply lib/Doctrine/DBAL/Query/QueryBuilder.php]: that's right

File: tests/Doctrine/Tests/DBAL/Query/QueryBuilderTest.php (hunk 2 above)
Comment (viosys) [reply tests/Doctrine/Tests/DBAL/Query/QueryBuilderTest.php]: to only add the columnnames to the insert statement

File: tests/Doctrine/Tests/DBAL/Query/QueryBuilderTest.php (hunk 1 above)
Comment (viosys) [reply tests/Doctrine/Tests/DBAL/Query/QueryBuilderTest.php]: that's right

File: lib/Doctrine/DBAL/Query/QueryBuilder.php (hunk 3 above)
Comment (Ocramius) [reply lib/Doctrine/DBAL/Query/QueryBuilder.php]: Throwing is a better idea here

File: lib/Doctrine/DBAL/Query/QueryBuilder.php (hunk 5 above)
Comment (Ocramius) [reply lib/Doctrine/DBAL/Query/QueryBuilder.php]: > slightly more complex statement

If you have such a scenario, please add a test for it

File: lib/Doctrine/DBAL/Query/QueryBuilder.php (hunk 4 above)
Comment (Ocramius) [reply lib/Doctrine/DBAL/Query/QueryBuilder.php]: the `! empty()` is sufficient

**Pull request closed without merging**